examples_apps/run_state_learner.py

```diff
@@ -0,0 +1,432 @@
+
+######################################################################
+# Now, we can create a function to define the :math:`i`\ th layer, acting
+# on qumode ``q``. This allows us to simply call this function in a loop
+# later on when we build our circuit.
+
+# layer architecture
+@operation(1)
+def layer(i, q):
+    Rgate(sf_params[i][0]) | q
+    Sgate(sf_params[i][1], sf_params[i][2]) | q
+    Rgate(sf_params[i][3]) | q
+    Dgate(sf_params[i][4], sf_params[i][5]) | q
+    Kgate(sf_params[i][6]) | q
+    return q
+
+
+######################################################################
+# Constructing the circuit
+# ------------------------
+#
+# Now that we have defined our gate parameters and our layer structure, we
+# can construct our variational quantum circuit.
+
+
+# Apply circuit of layers with corresponding depth
+with prog.context as q:
+    for k in range(depth):
+        layer(k) | q[0]
+
+
+######################################################################
+# Performing the optimization
+# ---------------------------
+#
+# :math:`\newcommand{ket}[1]{\left|#1\right\rangle}` With the Strawberry
+# Fields TensorFlow backend calculating the resulting state of the circuit
+# symbolically, we can use TensorFlow to optimize the gate parameters to
+# minimize the cost function we specify. With state learning, the measure
+# of distance between two quantum states is given by the fidelity of the
+# output state :math:`\ket{\psi}` with some target state
+# :math:`\ket{\psi_t}`. This is defined as the overlap between the two
+# states:
+#
+# .. math::  F = \left|\left\langle{\psi}\mid{\psi_t}\right\rangle\right|^2
+#
+# where the output state can be written
+# :math:`\ket{\psi}=U(\vec{\theta})\ket{\psi_0}`, with
+# :math:`U(\vec{\theta})` the unitary operation applied by the variational
+# quantum circuit, and :math:`\ket{\psi_0}=\ket{0}` the initial state.
+#
+# Let's define the target state as the single photon state
+# :math:`\ket{\psi_t}=\ket{1}`:
+
+import numpy as np
+
+target_state = np.zeros([cutoff])
+target_state[1] = 1
+print(target_state)
+
+
+######################################################################
+# Using this target state, we calculate the fidelity with the state
+# exiting the variational circuit. We must use TensorFlow functions to
+# manipulate this data, as well as a ``GradientTape`` to keep track of the
+# corresponding gradients!
+#
+# We choose the following cost function:
+#
+# .. math:: C(\vec{\theta}) = \left| \langle \psi_t \mid U(\vec{\theta})\mid 0\rangle - 1\right|
+#
+# By minimizing this cost function, the variational quantum circuit will
+# prepare a state with high fidelity to the target state.
+
+
+def cost(weights):
+    # Mapping from the Strawberry Fields free parameters to
+    # the TensorFlow variables.
+    mapping = {
+        p.name: w
+        for (sf_layer, weights_layer) in zip(sf_params, weights * 1.0)
+        for (p, w) in zip(sf_layer, weights_layer)
+    }
```

Comment:
I find this part quite hard to parse. It would be good to have a separate line explaining this and maybe printing out the dictionary when applied to `weights`?

Reply:
Agreed! Was `weights * 1.0` used as a shorthand for float conversion using `numpy`'s broadcasting?

Reply:
This is a weird bug/edge case in TensorFlow. While `tf.Tensor` objects _are_ iterables, `tf.Variable` objects aren't!

```python
>>> x = tf.Variable([0.0, 1.0, 2.0])
>>> list(x) # fails since Variables are not iterable
>>> list(x*1.0) # works; multiplying a Variable by a scalar casts it to a Tensor which *is* iterable
[<tf.Tensor: id=17, shape=(), dtype=int32, numpy=0>,
 <tf.Tensor: id=21, shape=(), dtype=int32, numpy=1>,
 <tf.Tensor: id=25, shape=(), dtype=int32, numpy=2>]
```

Reply:
@trbromley: I've updated this line to make it much simpler, I realised you don't need a nested dictionary comprehension if you just flatten both lists first! Check it out and let me know what you think.

Reply:
Great, looks much better thanks!

Reply:
Great, thank you! With the new version `tf.reshape(weights, [-1]))` would do the casting, right?

It might be worth capturing it in a comment: `# As TensorFlow Variables are not iterable, they are being cast to Tensors through reshaping`